### ISSUE TITLE
Crash when constraining view's top to snp_topLayoutGuideBottom on iOS9

### ISSUE THREAD

**Issue by TadeasKriz** [SnapKit/SnapKit#154]
```
2015-09-25 13:42:46.170 x-ios[40530:238886] The view hierarchy is not prepared for the constraint: <SnapKit.LayoutConstraint:0x7fef5db3f900@Classes/Controller/Registration/ProProfile/ProfileTypeSelectionController.swift#124 ProTypeView:0x7fef5da20150.top == _UILayoutSpacer:0x7fef5db5d5f0.bottom + 20.0>
	When added to a view, the constraint's items must be descendants of that view (or the view itself). This will crash if the constraint needs to be resolved before the view hierarchy is assembled. Break on -[UIView(UIConstraintBasedLayout) _viewHierarchyUnpreparedForConstraint:] to debug.
2015-09-25 13:42:46.189 x-ios[40530:238886] UILayoutGuide's owningView hierarchy unprepared for constraint.
	self = <_UILayoutSpacer: 0x7fef5db5d5f0 - "UIVC-topLayoutGuide", layoutFrame = {{0, 0}, {0, 0}}, owningView = <MainView: 0x7fef5d96b2d0; frame = (0 0; 375 667); autoresize = W+H; layer = <CALayer: 0x7fef5da14590>>>
	Constraint: <SnapKit.LayoutConstraint:0x7fef5db3f900@Classes/Controller/Registration/ProProfile/ProfileTypeSelectionController.swift#124 ProTypeView:0x7fef5da20150.top == _UILayoutSpacer:0x7fef5db5d5f0.bottom + 20.0>
	Container hierarchy: 
<StyledView: 0x7fef5da1e870; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5da8bbc0>>
   | <ProTypeView: 0x7fef5da20150; frame = (0 0; 0 0); gestureRecognizers = <NSArray: 0x7fef5a4beed0>; layer = <CALayer: 0x7fef5d9a8320>>
   |    | <RingIconView: 0x7fef5a4c8a40; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5a4c2b00>>
   |    |    | <StyledImageView: 0x7fef5a4c8d40; baseClass = UIImageView; frame = (0 0; 0 0); userInteractionEnabled = NO; layer = <CALayer: 0x7fef5a4c32a0>> - (null)
   |    |    | <InnerRing: 0x7fef5a488fb0; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5a4c2440>>
   |    | <UIImageView: 0x7fef5a4c8770; frame = (0 0; 0 0); opaque = NO; userInteractionEnabled = NO; layer = <CALayer: 0x7fef5a498e10>> - (null)
   |    | <StyledLabel: 0x7fef5a4c1890; baseClass = UILabel; frame = (0 0; 0 0); text = 'ARTIST'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5a4c2b70>>
   |    | <StyledLabel: 0x7fef5a4c7c30; baseClass = UILabel; frame = (0 0; 0 0); text = 'For bands and musicians! ...'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5a4c7e30>>
   | <ProTypeView: 0x7fef5da9f9e0; frame = (0 0; 0 0); gestureRecognizers = <NSArray: 0x7fef5daa0650>; layer = <CALayer: 0x7fef5da6a370>>
   |    | <RingIconView: 0x7fef5daa0680; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5daa0890>>
   |    |    | <StyledImageView: 0x7fef5daa0e40; baseClass = UIImageView; frame = (0 0; 0 0); userInteractionEnabled = NO; layer = <CALayer: 0x7fef5daa0ff0>> - (null)
   |    |    | <InnerRing: 0x7fef5daa10b0; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5daa1010>>
   |    | <UIImageView: 0x7fef5daa1830; frame = (0 0; 0 0); opaque = NO; userInteractionEnabled = NO; layer = <CALayer: 0x7fef5daa1710>> - (null)
   |    | <StyledLabel: 0x7fef5daa1a70; baseClass = UILabel; frame = (0 0; 0 0); text = 'PROMOTER'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5daa1c70>>
   |    | <StyledLabel: 0x7fef5daa26b0; baseClass = UILabel; frame = (0 0; 0 0); text = 'Tailor-made business tool...'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5daa28b0>>
   | <ProTypeView: 0x7fef5dab0d00; frame = (0 0; 0 0); hidden = YES; gestureRecognizers = <NSArray: 0x7fef5dab1910>; layer = <CALayer: 0x7fef5dab05a0>>
   |    | <RingIconView: 0x7fef5dab1940; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5dab1b50>>
   |    |    | <StyledImageView: 0x7fef5dab2140; baseClass = UIImageView; frame = (0 0; 0 0); userInteractionEnabled = NO; layer = <CALayer: 0x7fef5dab22f0>> - (null)
   |    |    | <InnerRing: 0x7fef5dab23b0; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5dab2310>>
   |    | <UIImageView: 0x7fef5dab2bb0; frame = (0 0; 0 0); opaque = NO; userInteractionEnabled = NO; layer = <CALayer: 0x7fef5dab2a90>> - (null)
   |    | <StyledLabel: 0x7fef5dab2df0; baseClass = UILabel; frame = (0 0; 0 0); text = 'VENUE'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5dab2ff0>>
   |    | <StyledLabel: 0x7fef5dab3a30; baseClass = UILabel; frame = (0 0; 0 0); text = 'For all venues, music clu...'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5dab3c30>>
   | <ProTypeView: 0x7fef5dac16e0; frame = (0 0; 0 0); hidden = YES; gestureRecognizers = <NSArray: 0x7fef5dac2340>; layer = <CALayer: 0x7fef5dac0fa0>>
   |    | <RingIconView: 0x7fef5dac2370; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5dac2580>>
   |    |    | <StyledImageView: 0x7fef5dac2b30; baseClass = UIImageView; frame = (0 0; 0 0); userInteractionEnabled = NO; layer = <CALayer: 0x7fef5dac2ce0>> - (null)
   |    |    | <InnerRing: 0x7fef5dac2da0; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5dac2d00>>
   |    | <UIImageView: 0x7fef5dac3520; frame = (0 0; 0 0); opaque = NO; userInteractionEnabled = NO; layer = <CALayer: 0x7fef5dac3400>> - (null)
   |    | <StyledLabel: 0x7fef5dac3760; baseClass = UILabel; frame = (0 0; 0 0); text = 'FESTIVAL'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5dac3960>>
   |    | <StyledLabel: 0x7fef5dac43a0; baseClass = UILabel; frame = (0 0; 0 0); text = 'For festivals of all genr...'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5dac45a0>>
   | <ProTypeView: 0x7fef5dac5840; frame = (0 0; 0 0); hidden = YES; gestureRecognizers = <NSArray: 0x7fef5a4ca4a0>; layer = <CALayer: 0x7fef5dac5700>>
   |    | <RingIconView: 0x7fef5a4ca4d0; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5a4ca6e0>>
   |    |    | <StyledImageView: 0x7fef5a4cac90; baseClass = UIImageView; frame = (0 0; 0 0); userInteractionEnabled = NO; layer = <CALayer: 0x7fef5a4cae40>> - (null)
   |    |    | <InnerRing: 0x7fef5a4caf00; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5a4cae60>>
   |    | <UIImageView: 0x7fef5a4cb680; frame = (0 0; 0 0); opaque = NO; userInteractionEnabled = NO; layer = <CALayer: 0x7fef5a4cb560>> - (null)
   |    | <StyledLabel: 0x7fef5a4cb8c0; baseClass = UILabel; frame = (0 0; 0 0); text = 'REGISTRATION.CHOOSEPROPRO...'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5a4cbac0>>
   |    | <StyledLabel: 0x7fef5a4cc850; baseClass = UILabel; frame = (0 0; 0 0); text = 'Registration.ChooseProPro...'; userInteractionEnabled = NO; layer = <_UILabelLayer: 0x7fef5a486df0>>
	View not found in container hierarchy: <MainView: 0x7fef5d96b2d0; frame = (0 0; 375 667); autoresize = W+H; layer = <CALayer: 0x7fef5da14590>>
	That view's superview: <_UIParallaxDimmingView: 0x7fef5db67fb0; frame = (0 0; 375 667); opaque = NO; layer = <CALayer: 0x7fef5db66cf0>>
2015-09-25 13:42:46.199 x-ios[40530:238886] *** Terminating app due to uncaught exception 'NSGenericException', reason: 'Unable to install constraint on view.  Does the constraint reference something from outside the subtree of the view?  That's illegal. constraint:<SnapKit.LayoutConstraint:0x7fef5db3f900@Classes/Controller/Registration/ProProfile/ProfileTypeSelectionController.swift#124 ProTypeView:0x7fef5da20150.top == _UILayoutSpacer:0x7fef5db5d5f0.bottom + 20.0> view:<StyledView: 0x7fef5da1e870; frame = (0 0; 0 0); layer = <CALayer: 0x7fef5da8bbc0>>'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000107c23f65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000109767deb objc_exception_throw + 48
	2   CoreFoundation                      0x0000000107c23e9d +[NSException raise:format:] + 205
	3   Foundation                          0x000000010933194c -[NSLayoutConstraint _addToEngine:integralizationAdjustment:mutuallyExclusiveConstraints:] + 187
	4   UIKit                               0x000000010a939213 __57-[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:]_block_invoke_2 + 504
	5   Foundation                          0x000000010933f23a -[NSISEngine withBehaviors:performModifications:] + 155
	6   UIKit                               0x000000010a939637 -[UIView(AdditionalLayoutSupport) _withAutomaticEngineOptimizationDisabledIfEngineExists:] + 58
	7   UIKit                               0x000000010a938fd3 __57-[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:]_block_invoke + 646
	8   UIKit                               0x000000010a939640 -[UIView(AdditionalLayoutSupport) _withAutomaticEngineOptimizationDisabledIfEngineExists:] + 67
	9   UIKit                               0x000000010a938d08 -[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:] + 242
	10  UIKit                               0x000000010a9390fc __57-[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:]_block_invoke_2 + 225
	11  Foundation                          0x000000010933f23a -[NSISEngine withBehaviors:performModifications:] + 155
	12  UIKit                               0x000000010a939637 -[UIView(AdditionalLayoutSupport) _withAutomaticEngineOptimizationDisabledIfEngineExists:] + 58
	13  UIKit                               0x000000010a938fd3 __57-[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:]_block_invoke + 646
	14  UIKit                               0x000000010a939640 -[UIView(AdditionalLayoutSupport) _withAutomaticEngineOptimizationDisabledIfEngineExists:] + 67
	15  UIKit                               0x000000010a938d08 -[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:] + 242
	16  UIKit                               0x000000010a172fef -[UIScrollView _switchToLayoutEngine:] + 107
	17  UIKit                               0x000000010a9390fc __57-[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:]_block_invoke_2 + 225
	18  Foundation                          0x000000010933f23a -[NSISEngine withBehaviors:performModifications:] + 155
	19  UIKit                               0x000000010a939637 -[UIView(AdditionalLayoutSupport) _withAutomaticEngineOptimizationDisabledIfEngineExists:] + 58
	20  UIKit                               0x000000010a938fd3 __57-[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:]_block_invoke + 646
	21  UIKit                               0x000000010a939640 -[UIView(AdditionalLayoutSupport) _withAutomaticEngineOptimizationDisabledIfEngineExists:] + 67
	22  UIKit                               0x000000010a938d08 -[UIView(AdditionalLayoutSupport) _switchToLayoutEngine:] + 242
	23  UIKit                               0x000000010a9388c2 -[UIView(AdditionalLayoutSupport) _initializeHostedLayoutEngine] + 649
	24  UIKit                               0x000000010a92ad8f -[UIView(UIConstraintBasedLayout) _layoutEngine_windowDidChange] + 183
	25  UIKit                               0x000000010a1628ac -[UIView(Internal) _didMoveFromWindow:toWindow:] + 216
	26  UIKit                               0x000000010a158106 __45-[UIView(Hierarchy) _postMovedFromSuperview:]_block_invoke + 155
	27  UIKit                               0x000000010a158005 -[UIView(Hierarchy) _postMovedFromSuperview:] + 544
	28  UIKit                               0x000000010a165ac3 -[UIView(Internal) _addSubview:positioned:relativeTo:] + 1967
	29  UIKit                               0x000000010a0a5840 -[_UIParallaxDimmingView didMoveToWindow] + 179
	30  UIKit                               0x000000010a162ebc -[UIView(Internal) _didMoveFromWindow:toWindow:] + 1768
	31  UIKit                               0x000000010a162b32 -[UIView(Internal) _didMoveFromWindow:toWindow:] + 862
	32  UIKit                               0x000000010a158106 __45-[UIView(Hierarchy) _postMovedFromSuperview:]_block_invoke + 155
	33  UIKit                               0x000000010a158005 -[UIView(Hierarchy) _postMovedFromSuperview:] + 544
	34  UIKit                               0x000000010a165ac3 -[UIView(Internal) _addSubview:positioned:relativeTo:] + 1967
	35  UIKit                               0x000000010a0a085a __53-[_UINavigationParallaxTransition animateTransition:]_block_invoke_2 + 2459
	36  UIKit                               0x000000010a15deec +[UIView(Animation) performWithoutAnimation:] + 65
	37  UIKit                               0x000000010a09fe60 __53-[_UINavigationParallaxTransition animateTransition:]_block_invoke + 258
	38  UIKit                               0x000000010a164082 +[UIView(Internal) _performBlockDelayingTriggeringResponderEvents:] + 188
	39  UIKit                               0x000000010a09f8b1 -[_UINavigationParallaxTransition animateTransition:] + 1432
	40  UIKit                               0x000000010a279d01 -[UINavigationController _startCustomTransition:] + 4112
	41  UIKit                               0x000000010a2887c7 -[UINavigationController _startDeferredTransitionIfNeeded:] + 712
	42  UIKit                               0x000000010a28967d -[UINavigationController __viewWillLayoutSubviews] + 57
	43  UIKit                               0x000000010a42163d -[UILayoutContainerView layoutSubviews] + 248
	44  UIKit                               0x000000011ca28aac -[UILayoutContainerViewAccessibility layoutSubviews] + 43
	45  UIKit                               0x000000010a16911c -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 710
	46  QuartzCore                          0x0000000109f7336a -[CALayer layoutSublayers] + 146
	47  QuartzCore                          0x0000000109f67bd0 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	48  QuartzCore                          0x0000000109f67a4e _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	49  QuartzCore                          0x0000000109f5c1d5 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 277
	50  QuartzCore                          0x0000000109f899f0 _ZN2CA11Transaction6commitEv + 508
	51  UIKit                               0x000000010a0b31b6 _UIApplicationHandleEventQueue + 7183
	52  CoreFoundation                      0x0000000107b500a1 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	53  CoreFoundation                      0x0000000107b45fcc __CFRunLoopDoSources0 + 556
	54  CoreFoundation                      0x0000000107b45483 __CFRunLoopRun + 867
	55  CoreFoundation                      0x0000000107b44e98 CFRunLoopRunSpecific + 488
	56  GraphicsServices                    0x000000010d4bfad2 GSEventRunModal + 161
	57  UIKit                               0x000000010a0b8676 UIApplicationMain + 171
	58  x-ios                         0x0000000106c241cd main + 109
	59  libdyld.dylib                       0x000000010c24692d start + 1
	60  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

And the code that creates the failing constraint is:
```
artist.snp_remakeConstraints { make in
    make.top.equalTo(snp_topLayoutGuideBottom).offset(MARGIN)
    make.leading.equalTo(contentView).offset(MARGIN)
    make.size.equalTo(size)
}
```

And it is in a controller that extends `BaseScrollViewController`:
```
@objc(BaseScrollViewController)
public class BaseScrollViewController: BaseViewController {
    public var scrollView: StyledScrollView!
    private var content: StyledView!
    
    public override var contentView: UIView {
        return content
    }
    
    public override func loadView() {
        super.loadView()
        scrollView => view
        content => scrollView
        addConstraints()
    }
    
    private func addConstraints() {
        scrollView.snp_makeConstraints { make in
            make.edges.equalTo(view)
        }
        
        content.snp_makeConstraints { make in
            make.top.bottom.equalTo(scrollView)
            make.left.right.equalTo(view)
        }
    }
}

@objc(BaseViewController)
public class BaseViewController: UIViewController, PostInitInjectable {
    public private(set) var navigator: Navigator!
    
    public var contentView: UIView {
        return view
    }
    
    public func inject(injector: Injector) {
        navigator <- injector
    }
    
    public override func loadView() {
        view = MainView()
    }
}
```

This works properly on iOS8. Also I tried to make it create the constraints in `updateViewConstraints`, but it still crashed.

**Comment by robertjpayne** [comment]
What version of SnapKit are you using?

**Comment by TadeasKriz** [comment]
I am on 0.15.0 release.

**Comment by robertjpayne** [comment]
@TadeasKriz it's a bit hard to follow your code, does it change at all though if you use `viewDidLoad` rather than `loadView`? Just looking at the error it appears to a be a view consistency thing now in iOS9 not a layout guide thing or SnapKit specific.

**Comment by TadeasKriz** [comment]
No it is the same. It has to be something with the layout guide, because when I make the top constraint to be constrained to its parent's top, it works properly. I'll check the iOS9 view consistency and what it is exactly.

**Comment by robertjpayne** [comment]
@TadeasKriz yea hard to say, layout guides are potentially more picky than regular ole views.

**Comment by TadeasKriz** [comment]
@robertjpayne It seems to be something with UIScrollView. Seems like you cannot constraint a view inside UIScrollView to something outside of it. I'll do more investigation on this issue though.

**Comment by robertjpayne** [comment]
@TadeasKriz shucks, yea probably nothing we can do about it here :(. Feel free to re-open or open a new issue if you find anything that lends itself to be a SnapKit specific issue.

**Comment by TadeasKriz** [comment]
I would like to reopen this issue because it can be indeed fixed in SnapKit. I tracked the issue to [this](https://github.com/SnapKit/SnapKit/blob/develop/Source/Constraint.swift#L318) line, where SnapKit calls `installOnView!.addConstraints(newLayoutConstraints)`. Documentation says that this will be deprecated in the future and that we should use `+[NSLayoutConstraint activateConstraints:]` instead. This method is available from iOS8+ and when I tried to replace the original line with `NSLayoutConstraint.activateConstraints(newLayoutConstraints)`, it started working properly with no crashes, yay!

What do you think? Could we add `if #available` there to call this method if we are on iOS8 and higher? This also applies to other methods dealing with constraints that are on UIView, as they have these equivalent methods in the `NSLayoutConstraint` class.

**Comment by robertjpayne** [comment]
@TadeasKriz the issue isn't really in SnapKit, we use `addConstraints` because we heuristically find out which view to add the constraints to where as `activateConstraints` does this automatically.

I'm guessing that something is now altered/broken in UIScrollView that our logic about finding which view to add the constraints to is invalid.

I'll dig in a bit and see where `activateConstraints` is adding them. Also the documentation has no deprecation warning or deprecation just says if you're developing for iOS 8 or later to use the `activateConstraints` method likely just because it solves a lot of issues people have in deciding which view the constraints need to be added to.

**Comment by robertjpayne** [comment]
Also I've edited the title of the bug to reflect the core issue at hand.

**Comment by TadeasKriz** [comment]
Oh, I did not know that. Thanks for the info. I guess my PR can be cancelled then.

**Comment by robertjpayne** [comment]
@TadeasKriz it may not be fixable otherwise but I'd like to check first.

**Comment by TadeasKriz** [comment]
@robertjpayne Also, it seems that constraining to a view outside of a UIScrollView is not the problem. I can constraint to an outside from an inside one just fine. The problem lies in constraining to the layout guide.

**Comment by robertjpayne** [comment]
@TadeasKriz ok, so sounds pretty specific from inside a UIScrollView to a layout guide on a view controller?

**Comment by TadeasKriz** [comment]
@robertjpayne That's right. I also tried to defer the creation of constraints (with a simple `dispatch_after` to run after 5 seconds of loadView) and it crashed as well, so it should not be a lifecycle problem.
